### PR TITLE
Sample: Use slashes instead of underscores in PHP namespaces for nested enums and messages

### PR DESCRIPTION
**Pull request by hzyi-google**

### CODE REVIEW

**Comment by hzyi-google** [comment]
PTAL

**Comment by hzyi-google** [comment]
We don't usually enforce that... but it does seem the changes in this PR could use some more testing. I'll see if I can reach the target code coverage.

**Comment by hzyi-google** [comment]
PTAL

**Comment by hzyi-google** [comment]
In this example, the second `Hello` will be used by its full name.
Added a test to show this logic too.

File: src/main/java/com/google/api/codegen/transformer/php/PhpModelTypeNameConverter.java [hunk 1]
```diff
@@ -153,67 +152,48 @@ public TypeName getTypeName(ProtoElement elem) {
...
+        builder.append('\\').append(CharMatcher.is('\\').trimFrom(namespace));
+        for (String component : components) {
           builder
-              .append('\\')
-              .append(components[index].substring(0, 1).toUpperCase())
-              .append(components[index].substring(1));
+              .append("\\")
+              .append(component.substring(0, 1).toUpperCase())
+              .append(component.substring(1));
+          return typeNameConverter.getTypeName(builder.toString());
```
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/php/PhpModelTypeNameConverter.java]: What is the point of this `for` loop if we always return on the first element?

**Comment by michaelbausor** [comment]
Fixes https://github.com/googleapis/gapic-generator/issues/2141

**Comment by hzyi-google** [comment]
@michaelbausor Thanks for the catch! Fixed and also added a test. PTAL

File: src/main/java/com/google/api/codegen/transformer/php/PhpModelTypeNameConverter.java [hunk 2]
```diff
@@ -153,67 +152,48 @@ public TypeName getTypeName(ProtoElement elem) {
...
-   *   <li>\Example\Top_Nested
-   *   <li>\Example\Top_Nested_DeepNested
+   *   <li>\Example\Top\Nested
+   *   <li>\Example\Top\Nested\DeepNested
    * </ul>
-   *
-   * <p>To correctly output these type names, we need to check whether the parent of a proto element
-   * is a message, and if so use '_' as a separator.
    */
   private TypeName getTypeName(ProtoElement elem, int maxDepth) {
```
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/php/PhpModelTypeNameConverter.java]: maxDepth is unused now - can we remove it?

File: src/main/java/com/google/api/codegen/transformer/php/PhpModelTypeNameConverter.java [hunk 3]
```diff
@@ -153,67 +152,48 @@ public TypeName getTypeName(ProtoElement elem) {
...
+        return typeNameConverter.getTypeName(builder.toString());
       }
     }
 
-    builder.append('\\').append(shortName);
-    return builder.toString();
+    // Create the type name based on the element's full name
+    for (String component : fullName.split("\\.")) {
+      builder
+          .append("\\")
```
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/php/PhpModelTypeNameConverter.java]: This step of splitting, appending `\`, capitalizing, and rejoining is duplicated. Consider the following refactor: create a helper function:

```java
private static String makeNamespacePiece(String piece) {
  return "\\" + piece.substring(0, 1).toUpperCase() + piece.substring(1);
}
```

Then use streams:

```java
String s = fullName.split("\\.").stream().transform(makeNamespacePiece).joinStrings();
```

(I don't think joinStrings() is a thing, but use a collector to join the strings in some way).

**Pull request merged**